apollo-router/src/services/subgraph_service.rs

```diff
@@ -917,39 +920,69 @@ pub(crate) async fn notify_batch_query(
             // graphql_response, so zip_eq shouldn't panic.
             // Use the tx to send a graphql_response message to each waiter.
             for (response, sender) in rs.into_iter().zip_eq(senders) {
-                sender
-                    .send(Ok(response))
-                    .map_err(|error| FetchError::SubrequestBatchingError {
-                        service: service.to_string(),
-                        reason: format!("tx send failed: {error:?}"),
-                    })?;
+                if let Err(log_error) =
+                    sender
+                        .send(Ok(response))
+                        .map_err(|error| FetchError::SubrequestBatchingError {
+                            service: service.to_string(),
+                            reason: format!("tx send failed: {error:?}"),
+                        })
+                {
+                    tracing::error!(service, error=%log_error, "failed to notify sender that batch processing succeeded");
+                }
             }
         }
     }
 
     Ok(())
 }
 
+type BatchInfo = (
+    (String, http::Request<Body>, Context, usize),
+    Vec<oneshot::Sender<Result<SubgraphResponse, BoxError>>>,
+);
+
 /// Collect all batch requests and process them concurrently
 #[instrument(skip_all)]
 pub(crate) async fn process_batches(
     client_factory: HttpClientServiceFactory,
     svc_map: HashMap<String, Vec<BatchQueryInfo>>,
 ) -> Result<(), BoxError> {
     // We need to strip out the senders so that we can work with them separately.
+    let mut errors = vec![];
     let (info, txs): (Vec<_>, Vec<_>) =
+    // let (info, txs) =
         futures::future::join_all(svc_map.into_iter().map(|(service, requests)| async {
-            let (_op_name, context, request, txs) = assemble_batch(requests).await;
+            let (_op_name, context, request, txs) = assemble_batch(requests).await?;
 
-            ((service, request, context, txs.len()), txs)
+            Ok(((service, request, context, txs.len()), txs))
         }))
         .await
         .into_iter()
+        .filter_map(|x: Result<BatchInfo, BoxError>| x.map_err(|e| errors.push(e)).ok())
```

Comment:
This is a super nit, but do we want this lambda to have side-effects?

Reply:
It's the only way I can think of to handle both errors and oks. Alternative approaches that achieve that are fine by me.